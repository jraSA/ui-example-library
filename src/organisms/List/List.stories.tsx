import React from 'react';
import { Meta } from '@storybook/react/types-6-0';
import List from '.';
import Text from '../../atoms/Text'

export default {
    title: 'Organisms/List',
    component: List,
} as Meta;

interface MyMockData {
    text: string
}

const myFooData: Array<MyMockData> = [
    { text: 'foo-1' },
    { text: 'foo-2' },
    { text: 'foo-3' }
];


const myBooData: Array<MyMockData> = [
    { text: 'boo-1' },
    { text: 'boo-2' },
    { text: 'boo-3' }
];

const renderItem = (item: MyMockData) => {
    return <Text>{item.text}</Text>;
}

const getData = () => {
    return myBooData;
}

const Template = (args) => <List {...args} />;

export const ListComponentWithoutData = Template.bind({});
ListComponentWithoutData.args = {
    renderItem: renderItem,
};

export const ListComponentWithIntialData = Template.bind({});
ListComponentWithIntialData.args = {
    renderItem: renderItem,
    initialData: myFooData
};

export const ListComponentWithGetData = Template.bind({});
ListComponentWithGetData.args = {
    renderItem: renderItem,
    getData: getData
};
