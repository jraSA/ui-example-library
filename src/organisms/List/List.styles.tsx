import styled from "styled-components";
import BaseTheme from "../../theme/BaseTheme";

const { space } = BaseTheme;

const Row = styled.div`
    margin-bottom: ${space[6]}px;
`;

export { Row };