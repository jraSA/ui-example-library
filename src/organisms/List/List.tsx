import React, { useEffect, useState } from 'react';
import Card from '../../molecules/Card';
import { Row } from './List.styles';

interface Props<T> {
    initialData?: T[],
    getData?: () => T[],
    renderItem: (item: T) => React.ReactNode
};

const List = <T extends object>(props: Props<T>) => {
    const { initialData, renderItem, getData } = props;
    const [data, setData] = useState(initialData)

    useEffect(() => {
        if (getData && !data) {
            const getNewData = async () => {
                const newData = await getData();
                setData(newData);
            }
            getNewData();
        }
    }, [data])

    const renderItems = () => {
        return (data ?
            data.map(item =>
                <Row key={JSON.stringify(item)}>
                    <Card>
                        {renderItem(item)}
                    </Card>
                </Row>
            ) :
            <div>loading!!</div>);
    };

    return (
        <ul>
            {renderItems()}
        </ul>
    );
};

export default List;
