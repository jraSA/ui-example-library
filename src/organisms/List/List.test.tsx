import React from 'react';
import renderer, { act } from 'react-test-renderer';
import List from '../List';
describe('List render test', () => {
    interface MyMockData {
        text: string
    }

    const myFooData: Array<MyMockData> = [
        { text: 'foo-1' },
        { text: 'foo-2' },
        { text: 'foo-3' }
    ];

    const myBooData: Array<MyMockData> = [
        { text: 'boo-1' },
        { text: 'boo-2' },
        { text: 'boo-3' }
    ];

    const renderItem = (item: MyMockData) => {
        return <div>{item.text}</div>
    }
    const getData = jest.fn();
    it('WHEN I dont provide data or getData THEN I should render loading', () => {
        const myList = renderer.create(
            <List renderItem={renderItem} />
        ).toJSON();
        expect(myList).toMatchSnapshot();
    });

    it('WHEN I provide initialData THEN I should render the list', async () => {
        const myList = renderer.create(
            <List initialData={myFooData} renderItem={renderItem} />
        ).toJSON();
        expect(myList).toMatchSnapshot();
    });

    it('WHEN I provide initialData and getData THEN I should render the list with the initialData', () => {
        const myList = renderer.create(
            <List initialData={myFooData} getData={getData} renderItem={renderItem} />
        ).toJSON();
        expect(myList).toMatchSnapshot();
    });

    it('WHEN I provide getData THEN I should render the list with the getData values', async () => {
        getData.mockReturnValueOnce(myBooData);

        let myList;
        await act(async () => {
            myList = renderer.create(
                <List getData={getData} renderItem={renderItem} />
            );
        });

        expect(getData).toBeCalled();
        expect(myList).toMatchSnapshot();
    });
});