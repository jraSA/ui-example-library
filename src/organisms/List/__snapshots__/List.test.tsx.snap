// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List render test WHEN I dont provide data or getData THEN I should render loading 1`] = `
<ul>
  <div>
    loading!!
  </div>
</ul>
`;

exports[`List render test WHEN I provide getData THEN I should render the list with the getData values 1`] = `
<ul>
  <div
    className="sc-eCssSg hkzTLf"
  >
    <div
      className="sc-gsTCUz Edwuc"
    >
      <div
        className="sc-bdfBwQ ciyLDR"
      >
        <div
          className="sc-hKgILt eoVmKB"
        >
          <div>
            boo-1
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="sc-eCssSg hkzTLf"
  >
    <div
      className="sc-gsTCUz Edwuc"
    >
      <div
        className="sc-bdfBwQ ciyLDR"
      >
        <div
          className="sc-hKgILt eoVmKB"
        >
          <div>
            boo-2
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="sc-eCssSg hkzTLf"
  >
    <div
      className="sc-gsTCUz Edwuc"
    >
      <div
        className="sc-bdfBwQ ciyLDR"
      >
        <div
          className="sc-hKgILt eoVmKB"
        >
          <div>
            boo-3
          </div>
        </div>
      </div>
    </div>
  </div>
</ul>
`;

exports[`List render test WHEN I provide initialData THEN I should render the list 1`] = `
<ul>
  <div
    className="sc-eCssSg hkzTLf"
  >
    <div
      className="sc-gsTCUz Edwuc"
    >
      <div
        className="sc-bdfBwQ ciyLDR"
      >
        <div
          className="sc-hKgILt eoVmKB"
        >
          <div>
            foo-1
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="sc-eCssSg hkzTLf"
  >
    <div
      className="sc-gsTCUz Edwuc"
    >
      <div
        className="sc-bdfBwQ ciyLDR"
      >
        <div
          className="sc-hKgILt eoVmKB"
        >
          <div>
            foo-2
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="sc-eCssSg hkzTLf"
  >
    <div
      className="sc-gsTCUz Edwuc"
    >
      <div
        className="sc-bdfBwQ ciyLDR"
      >
        <div
          className="sc-hKgILt eoVmKB"
        >
          <div>
            foo-3
          </div>
        </div>
      </div>
    </div>
  </div>
</ul>
`;

exports[`List render test WHEN I provide initialData and getData THEN I should render the list with the initialData 1`] = `
<ul>
  <div
    className="sc-eCssSg hkzTLf"
  >
    <div
      className="sc-gsTCUz Edwuc"
    >
      <div
        className="sc-bdfBwQ ciyLDR"
      >
        <div
          className="sc-hKgILt eoVmKB"
        >
          <div>
            foo-1
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="sc-eCssSg hkzTLf"
  >
    <div
      className="sc-gsTCUz Edwuc"
    >
      <div
        className="sc-bdfBwQ ciyLDR"
      >
        <div
          className="sc-hKgILt eoVmKB"
        >
          <div>
            foo-2
          </div>
        </div>
      </div>
    </div>
  </div>
  <div
    className="sc-eCssSg hkzTLf"
  >
    <div
      className="sc-gsTCUz Edwuc"
    >
      <div
        className="sc-bdfBwQ ciyLDR"
      >
        <div
          className="sc-hKgILt eoVmKB"
        >
          <div>
            foo-3
          </div>
        </div>
      </div>
    </div>
  </div>
</ul>
`;
